app/grandchallenge/codebuild/client.py

```diff
@@ -0,0 +1,99 @@
+import boto3
+from django.conf import settings
+from django.core import files
+from django.core.files.temp import NamedTemporaryFile
+
+from grandchallenge.algorithms.models import AlgorithmImage
+from grandchallenge.codebuild.models import Build
+from grandchallenge.core.storage import private_s3_storage
+
+
+class CodeBuildClient:
+    build = None
+
+    def __init__(
+        self, *, project_name=None, msg=None, algorithm=None, build_id=None,
+    ):
+        self.client = boto3.client(
+            "codebuild",
+            aws_access_key_id=settings.CODEBUILD_ACCESS_KEY,
+            aws_secret_access_key=settings.CODEBUILD_SECRET_KEY,
+            region_name=settings.CODEBUILD_REGION,
+        )
+        self.log_client = boto3.client(
+            "logs",
+            aws_access_key_id=settings.CODEBUILD_ACCESS_KEY,
+            aws_secret_access_key=settings.CODEBUILD_SECRET_KEY,
+            region_name=settings.CODEBUILD_REGION,
+        )
+        self.project_name = project_name
+        self.msg = msg
+        self.algorithm = algorithm
+        self.build_id = build_id
+        self.config = {
+            "name": project_name,
+            "source": {"type": "S3", "location": ""},
+            "artifacts": {
+                "type": "S3",
+                "location": settings.PRIVATE_S3_STORAGE_KWARGS["bucket_name"],
+            },
+            "environment": {
+                "type": "LINUX_CONTAINER",
+                "image": "docker:dind",
```

Comment:
Will we need a docker pro account to pull this? Should we instead have our own build images on ECR and pull from there?

Reply:
The number of pulls you can do as a free user is limited to (I think) 200 every 6 hours, but it adds up quickly. It would be nice if we can host our own for sure. For now we can at least use a docker hub account and feed the credentials into a secret in AWS Secrets Manager. The ARN for the secret can then be added under the `DOCKER_SECRET_ARN` setting.

Reply:
From https://docs.aws.amazon.com/codebuild/latest/userguide/sample-docker.html you should be able to use one of the standard images to do this

```
  "environment": {
    "type": "LINUX_CONTAINER",
    "image": "aws/codebuild/standard:4.0",
    "computeType": "BUILD_GENERAL1_SMALL",
    "environmentVariables": [
      {
        "name": "AWS_DEFAULT_REGION",
        "value": "region-ID"
      },
      {
        "name": "AWS_ACCOUNT_ID",
        "value": "account-ID"
      },
      {
        "name": "IMAGE_REPO_NAME",
        "value": "Amazon-ECR-repo-name"
      },
      {
        "name": "IMAGE_TAG",
        "value": "latest"
      }
    ],
    "privilegedMode": true
  },
```

Reply:
That would be nice. The the only bottleneck could be images pulled by the dockerfiles uploaded by users. I cannot add this change anymore though, since the config has been moved out of gc.